### ISSUE TITLE
DER: Dynamically Expandable Representation for Class Incremental Learning

### ISSUE THREAD

**Issue by HYUNMIN-HWANG** [One-Day-One-Paper/One-Day-One-Paper#306]
# DER: Dynamically Expandable Representation for Class Incremental Learning

![Untitled](https://user-images.githubusercontent.com/70581043/195281643-bb1e57f2-1342-4c18-9342-b15759666c65.png)

### [Overview]

1. Representation Learning Stage
    1. fix previous feature representation
    2. expand it with new feature extractor
    3. ausiliary loss on novel extractor
    4. channel-level mask-based pruning method
2. Classifier Learning Stage
    
    retrain the classifier with currently available data $\tilde{D}_t = D_t \bigcup M_t$ → class imbalance problem을 완화시켜준다.
    

### [Expandable Representation Learning]

step $t$, super-feature extractor $\Phi_t$, classifier $\mathcal{H}_t$, newly feature extractor $\mathcal{F_t}$

1. input x에서 feature extractor
    
    $u = \Phi_t(x) = [\Phi_{t-1}(x), \mathcal{F}_t(x)]$
    
    - freeze the learned funtion $\Phi_{t-1}$
    - Initialize $\mathcal{F}_t$
2. classification
    
    $p_{\mathcal{H}_t}(y|x)=Softmax(\mathcal{H}_t(u))$
    
3. predcition
    
    $\hat{y} = \mathrm{arg min} p_{\mathcal{H}_t}(y|x)$
    

Training Loss

- cross entropy loss
    
    $L_{\mathcal{H}_t} = - \frac{1}{|\tilde{D}_t|}\sum^{|\tilde{D}_t|}_{i=1} log(p_{\mathcal{H}_i}(y=y_i|x_i)))$
    
- auxiliary loss :  To enforce the network to learn the diverse and discriminative features for novel concepts
    
    $p_{H_t^\alpha}=Softmax(\mathcal{H}_t^\alpha(\mathcal{F}_t(x))$
    

### Channel-level Masks

differentiable channel-level masks

newly feature extractor $\mathcal{F_t}$, the input feature map of cnn layer $l$ as $f_l$, channel mask $m_l$

1. sigmoid : $m_l = \sigma(s{e_l})$
    - learnable mask parameters $e_l$
    - scaling factor to control the sharpness of the function $s$
        
        
![Untitled 1](https://user-images.githubusercontent.com/70581043/195281708-4588aa4c-5c67-4bd2-afea-d414d57ab54f.png)

        
2. channel mask와 cnn layer의 channel-level multiplication : $f_l'=f_l \odot m_l$
3. step t 에서의 super-feature : $\tilde{u} = \Phi_t^P(x)=[\mathcal{F}_1^p(x),\mathcal{F}_2^p(x), ..., \phi_t(x)]$ 

### Sparsity loss

based on the ration of used weights in all available weights.

$L_S=\frac{K_l||m_{l-1}||_1||m_l||_1}{\sum_{l=1}^LK_{lc_{l-1}c_l}}$

- $L$ : number of layers
- $K_l$  : kernel size of convolution layer $l$

=>> final loss : $L = L_{\mathcal{H}_t}+\lambda_a L_{\mathcal{H}_t^a}+\lambda_s L_S$

## Results
![Untitled 2](https://user-images.githubusercontent.com/70581043/195281679-fcb4d305-402d-4037-b9dd-9e72981c6f64.png)
